docs/guide/environmental_variables.rst

```diff
@@ -0,0 +1,71 @@
+Environmental Variables
+=======================
+
+Best practice holds that variables should be passed to graphs via environmental variables.
+Doing this is important for keeping sensitive data out of the code - such as an
+API token or username and password used to access a database. Not only is this
+approach more secure, it also makes graphs more flexible by allowing adjustments
+for a variety of environments and contexts. Importantly, environmental variables
+are also the means by-which arguments can be passed to graphs.
+
+
+Passing / Setting Environmental Variables
+::::::::::::::::::::::::::::::::::::::::::::
+
+The recommended way to set environmental variables for a given graph is simply to use
```

Comment:
I'm not completely at ease with this. I'd say it can be ok for now to word it like that, but I think the recommended way is to use environment variables, the way you set it depends on your context. You can set externally (for example, using a your shell, or your runner (docker, honcho, ...)), in the future you can set it using a .env file, and you can set it using arguments (--env, -e). The way you chose to set it depends on your constraints and environment, but all are strictly equivalent. Also, the "inheritance order" of those ways is the order shown here (I'll think today about this topic, but I think it make sense).

Reply:
Sorry, to match how other tools work, I think the precedence should be (first found is used):

* --env (bonobo ... --env FOO=bar)
* environment (FOO=bar bonobo ...)
* --env-file (bonobo ... --env-file=.env.local, not implemented yet)
* .env (bonobo ..., not implemented yet)

It makes more sense to read it this way, as it allows to use --env both for shells that does not support settings values on the fly, but also for overriding defaults provided in .env.

Let me know if you think I overlook something.

Reply:
I am a little confused about the precedence order. I thought that it worked in the opposite manner, that is, the last one set (or overridden) would be the one used. Therefore,

1. 'Global' = environment (FOO=bar bonobo ... or export FOO=bar; bonobo ...)
2. would think the env file would fall here as it would be used for graphs/modules run within the directory (not completely sure about the implementation) but wouldn't be set globally for the user or their shell session. ***(not yet implemented)
3. Setting or overriding environment variable = (--env FOO=bar or FOO=bar bonobo run...) would override both the global and .env file variables. 

So if I did the following, in order:

1. `export FIZZ=buzz`
2. In .env file - `FIZZ=fizz` or `export FIZZ=fizz`
3. `bonobo run --env FIZZ=fizzbuzz ...` or `FIZZ=fizzbuzz bonobo run ...`

FIZZ would be fizzbuzz for the graph because what counts is the last one set.  If the --env flag wasn't passed then FIZZ would be fizz for the graph because the .env file would override the value FIZZ was set-to at the global/shell/user scope. If neither 2 or 3 happened then FIZZ would be buzz for graph.

Am I misunderstanding something?

Regardless, I understand what you are saying and will make the changes you requested. After all, they can always be changed later.